File: experiment/websites/js/heapsort/heapsort.js
/* based on: https://rosettacode.org/wiki/Sorting_algorithms/Heapsort#JavaScript (2021-12-23) */

/**
 * @param {number[]} arr
 * @return {number[]}
 */
function heapsort(arr) {
  heapify(arr)
  end = arr.length - 1
  while (end > 0) {
    [arr[end], arr[0]] = [arr[0], arr[end]]
    end--
    siftDown(arr, 0, end)
  }
  return arr;
}

/**
 * @param {number[]} arr
 */
function heapify(arr) {
  start = Math.floor(arr.length/2) - 1

  while (start >= 0) {
    siftDown(arr, start, arr.length - 1)
    start--
  }
}

/**
 * @param {number[]} arr
 * @param {number} startPos
 * @param {number} endPos
 */
function siftDown(arr, startPos, endPos) {
  let rootPos = startPos

  while (rootPos * 2 + 1 <= endPos) {
    childPos = rootPos * 2 + 1
    if (childPos + 1 <= endPos && arr[childPos] < arr[childPos + 1]) {
      childPos++
    }
    if (arr[rootPos] < arr[childPos]) {
      [arr[rootPos], arr[childPos]] = [arr[childPos], arr[rootPos]]
      rootPos = childPos
    } else {
      return
    }
  }
}

File: experiment/websites/c/heapsort/heapsort.c
/* based on: https://rosettacode.org/wiki/Sorting_algorithms/Heapsort#C (2021-12-23) */

#include <emscripten/emscripten.h>

int max (int *a, int n, int i, int j, int k) {
  int m = i;
  if (j < n && a[j] > a[m]) {
    m = j;
  }
  if (k < n && a[k] > a[m]) {
    m = k;
  }
  return m;
}

void downheap (int *a, int n, int i) {
  while (1) {
    int j = max(a, n, i, 2 * i + 1, 2 * i + 2);
    if (j == i) {
      break;
    }
    int t = a[i];
    a[i] = a[j];
    a[j] = t;
    i = j;
  }
}

EMSCRIPTEN_KEEPALIVE
void heapsort (int *a, int n) {
  int i;
  for (i = (n - 2) / 2; i >= 0; i--) {
    downheap(a, n, i);
  }
  for (i = 0; i < n; i++) {
    int t = a[n - i - 1];
    a[n - i - 1] = a[0];
    a[0] = t;
    downheap(a, n - i - 1, 0);
  }
}
